numba/datamodel/arginfo.py

```diff
@@ -0,0 +1,118 @@
+from __future__ import print_function, absolute_import
+
+from collections import deque
+
+
+class ArgInfo(object):
+    """
+    Compute the position for each high-level typed argument.
+    It flattens every composite argument into primitive types.
```

Comment:
I am picking out the least common denominator here.  We used to pass structure into CUDA kernel by making a ctypes.Structure, allocating it in CUDA memory, transferring it over, pass the pointer to the device memory.  We can pass structures but we have to adjust it to match the alignment.  Getting the alignment right for nested structures are a lot of work.  Flattening is the simplest way.  Because:
* there are just a handful primitive scalar type.  So, fewer specialization.
* there are more conservative backends (e.g. OpenCL).  

>> Also, "ArgInfo" doesn't sound very descriptive :-)

Any suggestion to one of the hardest problem in CS.